samples/snap-geometry-edits-with-utility-network-rules/src/main/java/com/esri/arcgismaps/sample/snapgeometryeditswithutilitynetworkrules/components/SnapGeometryEditsWithUtilityNetworkRulesViewModel.kt

```diff
@@ -0,0 +1,514 @@
+    private val rulesLimitSymbol: Symbol =
+        SimpleLineSymbol(SimpleLineSymbolStyle.Solid, Color.fromRgba(255, 165, 0 ), 3f)
+    private val noneSymbol: Symbol =
+        SimpleLineSymbol(SimpleLineSymbolStyle.Dash, Color.green, 3f)
+    private val symbols = mutableMapOf(
+        SnapRuleBehavior.RulesPreventSnapping to rulesPreventSymbol,
+        SnapRuleBehavior.RulesLimitSnapping to rulesLimitSymbol,
+        SnapRuleBehavior.None to noneSymbol
+    )
+    private val symbolSwatches = mutableMapOf<SnapRuleBehavior, BitmapDrawable>()
+
+    // Save default renderers to reset the layers when a feature selection is cleared
+    private var defaultDistributionRenderer: Renderer? = null
+    private var defaultServiceRenderer: Renderer? = null
+
+    // Define a geometry editor with a snapping enabled. A reticle tool is ideal for touch devices.
+    val geometryEditor = GeometryEditor().apply {
+        snapSettings.isEnabled = true
+        snapSettings.isFeatureSnappingEnabled = true
+        tool = ReticleVertexTool()
+    }
+
+    // Define a graphics overlay which can also act as a snap source
+    private val defaultGraphicRenderer = SimpleRenderer(SimpleLineSymbol(
+        SimpleLineSymbolStyle.Dash, Color.fromRgba(165, 165, 165), 3f))
+    val graphicsOverlay = GraphicsOverlay(
+        graphics = listOf(Graphic(Geometry.fromJsonOrNull(application.getString(R.string.graphic_geometry_json))))
+    ).apply {
+        id = application.getString(R.string.graphics_overlay_id)
+        renderer = defaultGraphicRenderer
+    }
+
+    // Hold a reference to a selected feature and its related asset group and type
+    private var selectedFeature: ArcGISFeature? = null
+
+    private val _assetGroupNameState = MutableStateFlow("<Nothing selected>")
+    var assetGroupNameState: StateFlow<String> = _assetGroupNameState
+
+    private val _assetTypeNameState = MutableStateFlow("<Nothing selected>")
+    val assetTypeNameState: StateFlow<String> = _assetTypeNameState
+
+    // Represents the snap source settings object (for enabling and disabling), along with a name
+    // to use in the UI, and a symbol swatch
+    data class SnapSourceProperty(
+        val name: String,
+        val swatch: BitmapDrawable,
+        val snapSourceSettings: SnapSourceSettings
+    )
+    private val _snapSourcePropertyList = MutableStateFlow(listOf<SnapSourceProperty>())
+    internal val snapSourcePropertyList: StateFlow<List<SnapSourceProperty>> =
+        _snapSourcePropertyList
+
+    // Create boolean flags to track the state of UI components
+    private val _isEditButtonEnabled = MutableStateFlow(false)
+    internal val isEditButtonEnabled = _isEditButtonEnabled.asStateFlow()
+
+    init {
+        viewModelScope.launch {
+            arcGISMap.load().onFailure { error ->
+                messageDialogVM.showMessageDialog(
+                    "Failed to load map",
+                    error.message.toString()
+                )
```

Comment:
Could be simplified to use the error function: 
```suggestion
                messageDialogVM.showMessageDialog(error)
```